### PR TITLE
Feature/6: Edit habit route complete

### LINKED ISSUE #6: Habit Crud - Edit
As a user, I want to be able to edit  a habit goal so I can update my habit progress

### PR DESCRIPTION
**Pull request by sammychris**
Describe your changes here.

**Link to Issue**:

- Closes #6 

**Readiness**:

-   [x] Avoid duplicated logic
-   [x] Unit test component functionalities
-   [x] Use semantic HTML tags
-   [x] Use [semantic commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
-   [x] Connect this Pull Request to an issue

### CODE REVIEW

**Comment by Puritanic** [comment]
No tests?

**Comment by sammychris** [comment]
I am still working on it. But just want some code review first?

**Comment by sammychris** [comment]
Did the test.

File: routes/habitRoutes.js [hunk 1]
```diff
@@ -75,4 +75,22 @@ router.delete('/:id', passport.authenticate('jwt', { session: false }), (req, re
...
+ * @desc    Updates habit
+ * @access  Private
+ */
+router.put('/:id', passport.authenticate('jwt', { session: false }), (req, res) => {
+	const { name, description, type, difficulty, tags, frequency, user } = req.body;
+
+	// Check whether habit belongs to current authenticated user
+	Habit.findByIdAndUpdate(
+		{ _id: req.params.id },
+		{ $set: { name, description, type, difficulty, tags, frequency, user } },
```
Comment (Puritanic) [inline routes/habitRoutes.js]: I don't think that we want to update user, once Habit is created its author should be set in stone.

File: routes/habitRoutes.js [hunk 2]
```diff
@@ -75,4 +75,22 @@ router.delete('/:id', passport.authenticate('jwt', { session: false }), (req, re
...
+/**
+ * @route   PUT api/habit/:id
+ * @desc    Updates habit
+ * @access  Private
+ */
+router.put('/:id', passport.authenticate('jwt', { session: false }), (req, res) => {
+	const { name, description, type, difficulty, tags, frequency, user } = req.body;
+
+	// Check whether habit belongs to current authenticated user
+	Habit.findByIdAndUpdate(
```
Comment (Puritanic) [inline routes/habitRoutes.js]: This should be refactored to check for habit's user before updating it, I've explained that problem in Slack channel

File: routes/habitRoutes.js [hunk 3]
```diff
@@ -75,4 +75,22 @@ router.delete('/:id', passport.authenticate('jwt', { session: false }), (req, re
 		.catch(() => res.status(404).json({ message: 'Habit not found' }));
 });
 
+/**
+ * @route   PUT api/habit/:id
+ * @desc    Updates habit
+ * @access  Private
+ */
+router.put('/:id', passport.authenticate('jwt', { session: false }), (req, res) => {
+	const { name, description, type, difficulty, tags, frequency, user } = req.body;
```
Comment (Puritanic) [inline routes/habitRoutes.js]: I guess that we will need some sort of validation here too, just check if incoming values are satisfying requirements from createHabit validator.

File: __tests__/habits/habits.test.js [hunk 4]
```diff
@@ -205,4 +205,34 @@ describe('API - Habit', () => {
...
+				}));
+
+		test('If a valid habit ID is passed, that habit should be updated', done => {
+			request
+				.put(`/api/habit/${habitId}`)
+				.set('Authorization', token)
+				.end((err, res) => {
+					if (err) throw err;
+					expect(res.status).toBe(200);
+					expect(res.body.success).toBeTruthy();
```
Comment (Puritanic) [inline __tests__/habits/habits.test.js]: Maybe check if res.body[updated value] is actually updated here? Also, I'm not seeing any values that should be updated here, this kind of request shouldn't return success.

Comment (sammychris) [reply __tests__/habits/habits.test.js]: Thanks. It's done.

Comment (arielbk) [reply __tests__/habits/habits.test.js]: So this now updates the habit with the same values that it had before. Perhaps you could change some fields so it really is updating the habit.

There are still no tests to check any of the updated values of the habit that gets returned. So checking `res.body.habit.name`, etc.

File: routes/habitRoutes.js [hunk 5]
```diff
@@ -75,4 +75,37 @@ router.delete('/:id', passport.authenticate('jwt', { session: false }), (req, re
...
+			/* eslint-disable no-param-reassign */
+			habit.name = name;
+			habit.description = description;
+			habit.type = type;
+			habit.difficulty = difficulty;
+			habit.tags = tags;
+			habit.frequency = frequency;
+
+			/* eslint-enable */
+			return habit.save().then(() => res.status(200).json({ success: true, habit }));
```
[the remark this answers is not in the document]
Comment (sammychris) [reply routes/habitRoutes.js]: Thanks. You're right.

File: __tests__/habits/habits.test.js (hunk 4 above)
Comment (sammychris) [reply __tests__/habits/habits.test.js]: thanks. It skipped my mind.

File: routes/habitRoutes.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (sammychris) [reply routes/habitRoutes.js]: it done, thanks.

[the remark this answers is not in the document]
Comment (Puritanic) [reply routes/habitRoutes.js]: I think that @arielbk 	thought of something like this:
```
return habit.save().then((updatedHabit) => res.status(200).json({ success: true, habit: updatedHabit }));
```
You are still returning old modified habit.

File: bash.exe.stackdump [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Stack trace:
```
Comment (Puritanic) [inline bash.exe.stackdump]: What is this?

File: __tests__/habits/habits.test.js [hunk 7]
```diff
@@ -164,6 +169,51 @@ describe('API - Habit', () => {
...
+				.set('Authorization', token)
+				.send(habit)
+				.end((err, res) => {
+					if (err) throw err;
+					expect(res.status).toBe(404);
+					expect(res.body.message).toBe('Habit not found');
+					done();
+				}));
+
+		test('If a valid habit ID is passed, that habit should be updated', done => {
```
Comment (Puritanic) [inline __tests__/habits/habits.test.js]: Can you test something like this:
```
test('If the same habit ID is passed, it should no longer exist', done => {
			habit.name = 'New habit updated';

			request
 				.put(`/api/habit/${habitId}`)
 				.set('Authorization', token)
 				.send(habit)
 				.end((err, res) => {
 					if (err) throw err;
 					expect(res.status).toBe(200);
 					expect(res.body.habit.name).toBe('New habit updated');
					expect(res.body.habit. description).toBe(habit.description);
 					expect(res.body.success).toBeTruthy();
 					done();
 				});
		});
```
You should test if the value is updated, but also test if old values are not changed. Sorry if I'm annoying with this comments

File: bash.exe.stackdump (hunk 6 above)
Comment (sammychris) [reply bash.exe.stackdump]: It just came up, maybe because I am using window.

File: __tests__/habits/habits.test.js (hunk 7 above)
Comment (sammychris) [reply __tests__/habits/habits.test.js]: Okey. no problem.

File: bash.exe.stackdump (hunk 6 above)
Comment (Puritanic) [reply bash.exe.stackdump]: Put it .gitignore then, it's not project related.

File: __tests__/habits/habits.test.js (hunk 7 above)
Comment (sammychris) [reply __tests__/habits/habits.test.js]: any more changes? guys!!!

File: bash.exe.stackdump (hunk 6 above)
Comment (sammychris) [reply bash.exe.stackdump]: thanks just did.

**Comment by Puritanic** [comment]
LGTM

**Pull request merged**